### ISSUE TITLE
Add eval to template system

### ISSUE THREAD

**Issue by RhetTbull** [RhetTbull/osxphotos#417]
> 've already wondered if it wouldn't be useful to create more Template Substitutions, such as `{favorite?IS_FAV,NO_FAV}` or `{overall_score}`. But the above suggestion would be an elegant way of offering the same flexibility without creating more features.

Ahhm... I realized my assumption was wrong that `--quer-eval` was an elegant substitute for Template Substitutions. For instance, to have Templates in `--keyword`, `--filename` or `--directory` we'd need multiple runs of different `--query-eval`expressions, which may lead to unexpected results particularly with `--cleanup`, and might take a long time for a large library. 

Hence, it would be great to have more Templates like `{favorite}` and actually also `{place.name}`, in order to implement `--keyword` and `--filename` templates such as in `osxphotos export . --filename "{original_name}{edited?#,}{favorite?!,}" --keyword-template "{place.name}"`

Or maybe, how's this as a more flexible substitution syntax, to facilitate a wider variety of cases: A `=`as in `--filename "{original_name}{=photo.favorite?!,}"` or in `--keyword-template "{=place.name}"` could indicate that what follows the `=` is actually an **expression** similar to the query expressions in `--query-eval` — what do you think?

_Originally posted by @ubrandes in https://github.com/RhetTbull/osxphotos/issues/280#issuecomment-817194210_

**Comment by RhetTbull** [comment]
I think using the proposed `{=photo.field}` format would be difficult to support and maintain backwards compatibility with the current template syntax as the delimiter field currently occupies the space before the field name.   However, this could work like the `{exiftool:}` field does now in format:

`{eval:photo.favorite}` or drop the "photo" as it would always be implied: `{eval:favorite}` or, maybe `{photo:favorite}` and `{photo:place.name}` etc.

It would not be possible to fully support an eval as in `--query-eval` because of the current template syntax which reserves certain characters for special meaning, e.g. the `[find,replace]` syntax.

**Comment by RhetTbull** [comment]
If using `=` as proposed, the syntax would look like this when using a delimiter field:

`{,+=photo.albums}` as opposed to

`{,+eval:photo.albums}`

I think the second is easier to read.

**Comment by RhetTbull** [comment]
Actually, I think we could just use `photo.field` without any other template directive.  e.g. `{photo.favorite}`.  The template engine would detect the field name starts with `photo` then eval the statement in context of the current PhotoInfo object.

**Comment by ubrandes** [comment]
> detect the field name starts with photo then eval the statement in context of the current PhotoInfo object
> 

Sounds brilliant. `{photo.favorite}` and e.g. `{photo.place.name}` (or maybe even `{photo.score.overall>.7?!,}` etc?) seems so straightforward and powerful as a syntax.

Am 11.04.2021 um 01:28 schrieb Rhet Turnbull ***@***.***>:

**Comment by RhetTbull** [comment]
What's your use case for this feature (e.g. how would you use it)?  `--filename` and `--directory` only allow a single template so even with full eval capability, you'd still be limited in what you could evaluate in a single template string. 

I think this can be done fairly easily but I want to make sure I understand how you'd use it.

**Comment by ubrandes** [comment]
Thanks for asking. I specifically want to mark with letters # and/or ! between a photo’s filename and extension in my photo archive NAS if this photo was edited and/or favorite. 

I could imagine other use cases, such as marking
* photos from hometown,
* photos with a high overall score 
* photos with a certain label etc 
with either selected conditional keywords, and/or similar extra ASCII flags in their filename.

**Comment by RhetTbull** [comment]
For edited images you can accomplish this with `--edited-suffix`.  

I guess you could accomplish the others with the proposed eval template by chaining multiple templates each with a null default.  Eg:

`--filename "{original_name}{photo.favorite?!,}{photo.score.overall > 0.7?#,}"`

But finding text in a field would likely require a separate `eval:` template due to the way pythons's "in" keyword works. Eg:

`{eval:'beach' in photo.labels?beach,}`

**Comment by RhetTbull** [comment]
Any eval implementation will be limited though as the eval statement could not contain square brackets `[]` or parentheses `()` or curly braces `{}` as these all have special meanings in the osxphotos template language. Eg. This would not be valid: `{'beach' in photo.title.lower()?beach,}`

To really allow this kind of flexibility, a plug-in system would likely be better. Eg. Allow user to write a Python function that gets the PhotoInfo object for the photo and returns a string.

Eg:

`{plugin:plugin.py::myfunction}`

And plugin.py looks something like this:

```python 
def myfunction(photo):
    # your code here
    return replacement value

```

This lets you use all the logic / features of osxphotos export without having to write your own export code. You could use this anywhere a template is currently used.

**Comment by ubrandes** [comment]
Thank you, I wasn’t aware of --edited-suffix
(shoulda better RTFM ;)

Your `{photo.favorite?!,}` filename solution with the proposed {photo.<...>} template is exactly why I appreciate  your suggestion so much.

The separate `{eval:...}` template is not a super high priority for me, but others might see this differently. 

 Couldn’t Python‘s `in` operator be integrated, ie as if `{photo.labels?beach,}` effectively returned `{eval:'beach' in photo.labels?beach,}`?

**Comment by RhetTbull** [comment]
> Couldn’t Python‘s `in` operator be integrated, ie as if `{photo.labels?beach,}` effectively returned `{eval:'beach' in photo.labels?beach,}`?

No.  `?` is already used in the template system as `boolean?value if true,value if false`.  This is an edge case and I don't want to rewrite the entire template system (and break backwards compatibility) for this feature.   If you want to do anything complex for naming output files, the proposed plugin system will be the easiest way to do this.  

I *may* be able to implement something like the following to extend the template system with simple logic. And this would apply to all templates, not just `photo.property` template types:

`{template_field [contains|<|>] value}`

e.g.:

`{photo.score.overall < 0.7?!,}`

or
`{photo.persons contains 'John'?john,}`

`contains` would work for list values (like persons above) or string values (like `photo.title`) and would effectively implement python's `in`.

In order of implementation, I will likely do the following:

1. Add `photo` template to provide access to underlying PhotoInfo object
2. Add `plugin` (or possibly `function`) template to allow calling custom python function inside a template
3. Add the eval extension to allow `<`, `>`, and `contains`.  

I'll save the latter for last as this requires modifying the template grammar which takes a fair amount of testing.  The first two should be fairly easy to implement.

**Comment by RhetTbull** [comment]
Should also add `!=`

**Comment by ubrandes** [comment]
I’m glad to hear the photo template to provide access to underlying PhotoInfo object should be fairly easy to implement!

Re. use of `?`, you’re absolutely right, of course.  Obviously, anything after the `?` already states `<value if true, value if false>`.  So indeed, any boolean expression needs to stated be before the `?`   Your suggestions with `</>/contains` makes a lot of sense.

**Comment by RhetTbull** [comment]
I've added `{favorite}` as well as `{photo}` template as described above to [v0.41.11](https://github.com/RhetTbull/osxphotos/releases/tag/v0.41.11)

**Comment by ubrandes** [comment]
Awsome, thanks!

**Comment by RhetTbull** [comment]
I have a prototype of conditional expressions working.  Need to add tests but so far it seems very robust.  It will work like this:

`{template_field operator value?value_if_true,value_if_false}`

the valid operators so far are:

- `contains`
- `==`
- `!=`
- `<`
- `>`


Still need to add `<=`,`>=`

The value can itself be another full template statement so the following would be valid:
`{title == {title},OF COURSE!,}`

You can use filters to make case-insensitive matching:
`{keyword|lower contains england?England,NotEngland}`

This looks like it'll be quite powerful and will of course work with every template field, not just the new `{photo}` field.

**Comment by RhetTbull** [comment]
Wonder if I should add a `!contains`?  The same can be achieved using the `value_if_false` for `contains` but might might be easier to use if there's an explicit `not` operator?

**Comment by RhetTbull** [comment]
values in the template system can be a single value (e.g. `photo.title` which is a single string) or a list of values (e.g. `photo.keywords`).  The conditional logic needs to "do the right thing" in both cases.  For `<`, `>`, `<=`, `>=` I think the right thing is to operate only on single values and only if both values can be converted to numbers (float).  E.g. `{photo.score.overall < 0.7}` is valid but `{photo.title < 0.7}` wouldn't make sense and would be invalid.

For `==` and `!=`, the current implementation compares two scalar values if both sides of the expression are single values otherwise it compares sorted lists.  Thus, these are both valid expressions:

`{photo.title == Day at the beach}` (note that words are not quoted, this is by design to simplify the parsing)
`{photo.keywords == {exiftool:IPTC:Keywords}}` (True if Photos' keywords matches the EXIF keywords in the file, regardless of what order they're in)

For `contains`, "the right thing" is somewhat more complex.  For example, this would be valid:

`{photo.title contains Beach}` or, ignoring case, `{photo.title|lower contains beach}`

But what to do if a keyword is "BeachDay" and the expression is:
`{keyword contains Beach}`

In this case, should Beach match BeachDay (Beach is contained as a substring of BeachDay) or not because "Beach" is not one of the keywords in the list ["BeachDay", ...]?  I think the "right thing" is to match in this case because doing one thing when the value is a scalar and a different thing when the value is a list is not intuitive (and the way the template engine works, would require a lot of to make this happen).  

So if `contains` is implemented as described above, what to do if user wants to match a list exactly?  e.g. match if keyword contains `Beach` but not `BeachDay`?

I think the cleanest solution is a separate operator that means "match only if the item is in list".  Perhaps `matches`? as in:

```
keywords = ["BeachDay", "Travel", "Vacation"]
{keyword contains Beach} --> True
{keyword matches Beach} --> False
{keyword contains BeachDay} --> True
{keyword matches BeachDay} --> False
```

**Comment by ubrandes** [comment]
!contains ﻿sounds consistent — maybe it would round it off?

Also your reflections around „contains“ and „matches“ seem very consistent and considerate to me. I’m just getting sensitive of scope increase here: One might also wish to add „begins“ and „ends“ for strings beginning and ending with a phrase. Plus the !not versions... 

Maybe it’s easier to introduce a |not pipe similar to |lower and |upper that’s generic for Booleans, instead of introducing lots of !operators. Personally I like the brevity of ! but |not might also have merits.

**Comment by RhetTbull** [comment]
Thanks for the input!

> One might also wish to add „begins“ and „ends“ for strings beginning and ending with a phrase.

I've already added those! `startswith`, `endswith` (to be consistent with python's string methods of same name)

> Maybe it’s easier to introduce a |not pipe similar to |lower and |upper that’s generic for Booleans, 

the filters like `|lower` get evaluated *before* the conditional expressions so this wouldn't work unless added to the end of the conditional expression which doesn't help readability.

I'm debating between using '!' or 'not' for negation:

`{keyword not contains 'Beach'}`
vs
`{keyword !contains 'Beach'}`

Personally, while I like the succinctness of `!`, I think the `not` is more clear and likely easier for those who aren't experienced programmers to understand and use.  If implemented, 'not' would apply to all operators so a user could do this (and it would work as expected):

`{photo.score.overall not < 0.7}` 
vs
`{photo.score.overall >= 0.7}`

**Comment by RhetTbull** [comment]
The current implementation also allows multiple values to be used for the comparison value, separated by `|`.  This is very flexible but also means that `|` can't be used as a comparison value.  The current template engine doesn't support escaping of characters, e.g. `\|`.

e.g.
`{keyword matches Travel|Vacation}`

**Comment by RhetTbull** [comment]
Implemented in [v0.42.00](https://github.com/RhetTbull/osxphotos/releases/tag/v0.42.00)

If you use this feature and have any feedback, let me know!  For docs see `osxphotos help export` or [here](https://github.com/RhetTbull/osxphotos#command-line-reference-export)
